### PR TITLE
custom message and status codes in pass check callback

### PR DESCRIPTION
**Pull request by celesteking**
Which is basically an enhanced #976.

Magically, it also makes this code possible:
```
cb(false, {message: DSN.no_such_user()});
```

### CODE REVIEW

File: plugins/auth/auth_base.js [hunk 1]
```diff
@@ -83,12 +83,19 @@ exports.check_user = function (next, connection, credentials, method) {
         return;
     }
 
-    var passwd_ok = function (valid, message) {
+    // otps: {
```
Comment (msimerson) [inline plugins/auth/auth_base.js]: I'm thinking you meant, "opts"

**Comment by msimerson** [comment]
Besides the comment typo, LGTM.

**Comment by msimerson** [comment]
Note that this isn't backwards compatible with @DarkSorrow patch. As this PR would break his code, it'd be nice for either:
1. him to chime in here and bless the change (and update his code)
2. for this to check `typeof opts`, and use it as the message if it's a string

**Comment by smfreegard** [comment]
I'd prefer option 2 as it's more how Haraka works internally with the DSN module and is more developer friendly e.g. pass a string or a DSN object and it works it out automatically.

In a future Haraka major release; we can change *all* of the error messages internally and in all of the plugins to use enhanced status codes and then advertise ENHANCEDSTATUSCODES as an ESMTP extension.

**Comment by celesteking** [comment]
Here you go. (I don't know how to write tests yet)

File: plugins/auth/auth_base.js [hunk 2]
```diff
@@ -83,12 +83,18 @@ exports.check_user = function (next, connection, credentials, method) {
         return;
     }
 
-    var passwd_ok = function (valid, message) {
+    // valid: (true|false)
+    // opts: ({ message, code }|String)
+    var passwd_ok = function (valid, opts) {
```
Comment (smfreegard) [inline plugins/auth/auth_base.js]: You should use === over == as it's stricter and we can also ensure that we've got a DSN object too, additionally if we aren't using the DSN module, we shouldn't return any enhanced codes.

I would do it this way:

`````javascript
// valid:  true/false
// message:  optional response string or DSN object
var passwd_ok = function (valid, message) {
    var code, reply;
    if (typeof(message) === 'object' && message.constructor.name === 'DSN') {
        code = message.code;
        reply = message.reply;
    }
    else {
        code = (valid ? 235 : 535);
        reply = (message ? message : (valid ? 'Authentication successful' : 
                                              'Authentication failed'));
    }

    ....
    connection.respond(code, reply, function () ....
`````

**Comment by smfreegard** [comment]
Crap - I've just realised.   This whole patch is unnecessary.
See https://github.com/baudehlo/Haraka/blob/master/connection.js#L463

I already handled this in the respond() function....

**Comment by celesteking** [comment]
I noticed that first, when trying to understand why the heck status code was skipped when DSN obj was supplied. 

I need to accept both message & code because DSN doesn't support x.7.8, and I found no option to craft custom extended codes there... So that's why I can do the following in auth plugin: 
```return cb(success, {message: '5.7.8 Login or password incorrect', code: 535});```
and it "just works"

**Comment by smfreegard** [comment]
It's pretty simple:

`````
var DSN = require('dsn');
...
return cb(success, return new DSN(535, 'Login or password incorrect', 535, 7, 8));
`````

The DSN module was the very first thing I wrote for Haraka, and it would seem that I didn't include RFC4594 codes to it; should be pretty simple to add though.

**Comment by celesteking** [comment]
no, it's not simple.

```
> var DSN =  require('Haraka/dsn');
> new DSN(535, 'Login or password incorrect', 535, 7, 8);
TypeError: object is not a function
```

**Comment by smfreegard** [comment]
Ok - see https://github.com/baudehlo/Haraka/pull/985 which fixes that:

`````javascript
[smf@mail1-ec2 Haraka]$ node 
> var DSN = require('./dsn');
undefined
> new DSN.DSN(535, 'Login or password incorrect', 535, 7, 8);
{ code: 535,
  msg: 'Login or password incorrect',
  cls: 5,
  sub: 7,
  det: 8,
  default_msg: '',
  reply: '5.7.8 Login or password incorrect' }
`````

**Comment by smfreegard** [comment]
Fixed in #991 closing.

**Comment by celesteking** [comment]
this is work in progress actually i guess

**Pull request closed without merging**